client/hotrod/src/main/java/org/infinispan/hotrod/impl/cache/RemoteCacheImpl.java

```diff
@@ -219,12 +223,22 @@ public CompletionStage<Boolean> replace(K key, V value, CacheEntryVersion versio
       ReplaceIfUnmodifiedOperation<K, V> op = cacheOperationsFactory.newReplaceIfUnmodifiedOperation(key, keyToBytes(key), valueToBytes(value),
             ((CacheEntryVersionImpl) version).version(), options, dataFormat);
       // TODO: add new op to prevent requiring return value?
-      return op.execute().thenApply(r -> r.getValue() != null);
+      return op.execute().thenApply(r -> r.getCode().isUpdated());
    }
 
    @Override
    public CompletionStage<CacheEntry<K, V>> getOrReplaceEntry(K key, V value, CacheEntryVersion version, CacheWriteOptions options) {
-      throw new UnsupportedOperationException();
+      if (!(Objects.requireNonNull(version) instanceof CacheEntryVersionImpl)) {
+         throw new IllegalArgumentException("Only CacheEntryVersionImpl instances are supported!");
+      }
+      ReplaceIfUnmodifiedOperation<K, V> op = cacheOperationsFactory.newReplaceIfUnmodifiedOperation(key, keyToBytes(key), valueToBytes(value),
```

Comment:
Good catch, yeah we need to do that for single key operations otherwise it could go to the wrong node.

Reply:
This actually reminds me looking at getAll that it is not working properly with object storage. I can log a separate JIRA for that though as it will take more work.

Reply:
I created https://issues.redhat.com/browse/ISPN-14006 to handle the putAll at a later point. TBH it might be only worth fixing in the new client as object storage is not that common and it only causes additional latency.